c++/src/kj/compat/http.c++

```diff
@@ -2594,12 +2663,167 @@ private:
+          }
+          result = deflateInit2(
+              &ctx,
+              Z_DEFAULT_COMPRESSION,
+              Z_DEFLATED,
+              windowBits,
+              8,  // memLevel = 8 is the default
+              Z_DEFAULT_STRATEGY);
+          break;
+        case Mode::DECOMPRESS:
+          windowBits = -config.inboundMaxWindowBits.orDefault(15);
+          // We use negative values because we want to use raw inflate.
+          result = inflateInit2(&ctx, windowBits);
+          break;
+        default:
+          KJ_UNREACHABLE;
+      }
+      KJ_REQUIRE(result == Z_OK, "Failed to initialize compression context.");
+    }
+
+    ~Context() noexcept(false) {
+      switch (mode) {
+        case Mode::COMPRESS:
+          deflateEnd(&ctx);
+          break;
+        case Mode::DECOMPRESS:
+          inflateEnd(&ctx);
+          break;
+      }
+    }
+
+    KJ_DISALLOW_COPY(Context);
+
+    void setInput(const void* in, size_t size) {
+      ctx.next_in = const_cast<byte*>(reinterpret_cast<const byte*>(in));
+      ctx.avail_in = size;
+    }
+
+    Result pumpOnce() {
+      ctx.next_out = buffer;
+      ctx.avail_out = sizeof(buffer);
+
+      int result = Z_OK;
+
+      switch (mode) {
+        case Mode::COMPRESS:
+          result = deflate(&ctx, Z_SYNC_FLUSH);
+          KJ_REQUIRE(result == Z_OK || result == Z_BUF_ERROR || result == Z_STREAM_END,
+                      "Compression failed.");
+          break;
+        case Mode::DECOMPRESS:
+          result = inflate(&ctx, Z_SYNC_FLUSH);
+          KJ_REQUIRE(result == Z_OK || result == Z_BUF_ERROR || result == Z_STREAM_END,
+                      "Decompression failed.");
+          break;
+        default:
+          KJ_UNREACHABLE;
+      }
+
+      return Result {
+        .processResult = result,
+        .buffer = kj::arrayPtr(buffer, sizeof(buffer) - ctx.avail_out),
+      };
+    }
+
+    kj::OneOf<kj::Vector<kj::byte>, kj::Exception> processMessage(kj::ArrayPtr<const byte> message) {
+      // If `this` is the compressor, calling `processMessage()` will compress the `message`.
+      // Likewise, if `this` is the decompressor, `processMessage()` will decompress the `message`.
+      // If (de)compression is successful, the result is returned as a Vector, otherwise,
+      // an Exception is returned.
+
+      setInput(message.begin(), message.size());
+
+      kj::Vector<kj::byte> result;
+      KJ_IF_MAYBE(e, processLoop(result)) {
+        return kj::mv(*e);
+      }
+      return kj::mv(result);
+    }
+
+    kj::Maybe<kj::Exception> processLoop(kj::Vector<kj::byte>& output) {
+      // Since Zlib buffers the writes, we want to continue processing until there's nothing left.
+      Result result;
+      KJ_IF_MAYBE(exception, kj::runCatchingExceptions([this, &result]() {
+        result = pumpOnce();
+      })) {
+        return kj::mv(*exception);
+      }
+
+      if (result.buffer.size() > 0) {
+        output.addAll(result.buffer);
```

Comment:
We will need to bail out when the total number of decompressed bytes exceeds the message size limit. Otherwise, it's possible to construct a malicious message whose compressed size is very small, but which uncompresses to a huge amount of data, consuming all available memory.

Reply:
https://github.com/capnproto/capnproto/pull/1505/commits/2188f68f065b2eceaf2b52ff748056fa4e72d3fd

Reply:
Hmm, on second thought this looks wrong. If the client splits the message into `n` frames, then each time we add a fragment and call `receive()` we're passing a smaller `maxSize` until we get the final frame. At this point, we start decompressing the message, but `maxSize` isn't the original `maxSize`, it just tells us how large the remaining payload can be :thinking:. Will need to rethink this.

Reply:
I think this is a fairly non-invasive way to get back the original maximum size: https://github.com/capnproto/capnproto/pull/1505/commits/3eaf8d6672c45ab1ea11b72edbaa50c8f7141724